### ISSUE TITLE
Blur placeholder of `next/image` is still visible after the image is loaded

### ISSUE THREAD

**Issue by mmazzarolo** [vercel/next.js#53329]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.6.0
      npm: 8.13.2
      Yarn: 1.22.18
      pnpm: 8.6.7
    Relevant Packages:
      next: 13.4.13-canary.6
      eslint-config-next: 13.4.9
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Image optimization (next/image, next/legacy/image)

### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/pensive-mccarthy-yfz9lg

### To Reproduce

- Use `next/image` to load an image that has a transparent background.
- Set a `placeholder="blur"` on such image.
- Check your app: the image blur won't be removed immediately after the image has been loaded. This makes the blurry image visible underneath the loaded image (since it has a transparent background). 

### Describe the Bug

The image/style generated by setting `placeholder="blur"` on a `next/image` can still be seen (for some time) after the image has been loaded.  

Example from the attached Codesandbox:
<img width="908" alt="ScreenShot 2023-07-29 at 11 12 30@2x" src="https://github.com/vercel/next.js/assets/9536354/92a7ff0f-60f8-4727-a954-0f9f1055743b">

You can see other people reporting and discussing this bug at https://github.com/vercel/next.js/issues/42140#issuecomment-1342533513 .

For reference, here's how a Gatsby image with blur setup looks like out-of-the box:

https://github.com/vercel/next.js/assets/9536354/3484be15-066e-4ad0-9eef-4fd7c0c238a0

### Expected Behavior

The placeholder image/style should immediately disappear when the image is loaded. Or, the generated image placeholder should respect the transparency pixels of the used image. 

### Which browser are you using? (if relevant)

`Chromium Engine Version 115.0.5790.114`, `Safari Version 16.5.2 (18615.2.9.11.10)`

<sub>[NEXT-1679](https://linear.app/vercel/issue/NEXT-1679/blur-placeholder-of-nextimage-is-still-visible-after-the-image-is)</sub>

**Comment by Jordaneisenburger** [comment]
Yep, I'm experiencing the same issue when using `placeholder="blur"` and plaiceholder for for base64Data

**Comment by focux** [comment]
Same, I'm having this same issue. Have any of you found a work-around?

**Comment by Jorundur** [comment]
@focux A workaround is using the `onLoad` function on the Image component.

You can have a state variable to keep track of whether or not the image has loaded and only add the blur props when the image hasn't finished loading, i.e.

```ts
  ...
  const [imageLoaded, setImageLoaded] = React.useState<boolean>(false);

  return (
    <Image
      src={image.url}
      ...someProps
      {...(blurDataURL && !imageLoaded && { placeholder: 'blur', blurDataURL })}
      onLoad={() => {
        setImageLoaded(true);
      }}
    />
  );
};

```

**Comment by styfle** [comment]
This bug only impacts `next dev` with App Router.

When I try with `next build && next start`, then the blur placeholder is removed as soon as the image is loaded.

**Comment by ackvf** [comment]
Nope, this doesn't work, the `onLoad` will not fire until the image 100% finishes loading.

![gif-placeholder-bug-fast](https://github.com/vercel/next.js/assets/2313018/daa1f3ce-877f-48eb-9603-f6757f09b7ff)


See here, it's as if the gif was still loading frame by frame (hence why the first animation is so slow), so technically the image hasn't finished loading. But it is already being displayed (albeit frame by frame) at which point I don't want to see the placeholder. Second animation goes on full speed and the placeholder is no longer visible.

**Comment by ackvf** [comment]
Alright, this works like a charm! Just use it in place of Next/Image. It has the same signature.

Before:
```tsx
          <Image
            src={getStaticAssetURL('core.gif')}
            placeholder={coreStaticDataURL}
          />
```

After:
```tsx
          <LoadingImage
            src={getStaticAssetURL('core.gif')}
            placeholder={coreStaticDataURL}
          />
```
---

```tsx
/* eslint-disable jsx-a11y/alt-text */

import Image from 'next/image'
import { useReducer } from 'react'

//@ts-ignore
export const LoadingImage: React.FC<LoadingImageProps> = ({ placeholder, ...props }) => {
  //@ts-ignore
  const [isLoading, stopLoading] = useReducer((_, ev) => (props.onLoad?.(ev), false), true)

  return <>
    {isLoading && placeholder && <Image
      {...props}
      placeholder={undefined}
      src={placeholder}
    />}
    <Image
      {...props}
      className={`${props.className} ${isLoading && 'invisible'}`}
      placeholder={undefined}
      onLoad={stopLoading}
    />
  </>
}

export type NextImageProps = Parameters<typeof Image>[0]

export interface LoadingImageProps extends NextImageProps {
  placeholder: NextImageProps['placeholder']
}
```
You'll need to define also css
```
.invisible {
  visibility: hidden;
}
```

**Comment by artemshchirov** [comment]
> Alright, this works like a charm! Just use it in place of Next/Image. It has the same signature.
> 
> Before:
> 
> ```tsx
>           <Image
>             src={getStaticAssetURL('core.gif')}
>             placeholder={coreStaticDataURL}
>           />
> ```
> 
> After:
> 
> ```tsx
>           <LoadingImage
>             src={getStaticAssetURL('core.gif')}
>             placeholder={coreStaticDataURL}
>           />
> ```
> 
> ```tsx
> /* eslint-disable jsx-a11y/alt-text */
> 
> import Image from 'next/image'
> import { useReducer } from 'react'
> 
> /**
>  * This component is a drop-in replacement for next/image that shows a placeholder image while the image is loading.
>  * 
>  * *GitHub Issue: Placeholder of next/image is still visible after the image is loaded*\
>  * https://github.com/vercel/next.js/issues/53329
>  */
> 
> //@ts-ignore
> export const LoadingImage: React.FC<LoadingImageProps> = ({ placeholder, ...props }) => {
>   //@ts-ignore
>   const [isLoading, stopLoading] = useReducer((_, ev) => (props.onLoad?.(ev), false), true)
> 
>   // prettier-ignore
>   return (
>     <>
>       {isLoading && placeholder &&
>         <Image
>           {...props}
>           placeholder={undefined}
>           src={placeholder}
>         />
>       }
>       <Image
>         {...props}
>         className={`${props.className} ${isLoading && 'absolute invisible'}`}
>         placeholder={undefined}
>         onLoad={stopLoading}
>       />
>     </>
>   )
> }
> 
> export type NextImageProps = Parameters<typeof Image>[0]
> 
> export interface LoadingImageProps extends NextImageProps {
>   placeholder: NextImageProps['placeholder']
> }
> ```
> 
> You'll need to define also css (uses tailwind)
> 
> ```
> .invisible {
>   visibility: hidden;
> }
> .absolute{
>   position: absolute;
> }
> ```

Thank you!

**Comment by artemshchirov** [comment]
@ackvf Thank you!

**Comment by thanhtutzaw** [comment]
I also get this issue in nextjs 14 and blur data did not have border radius from original photo .

**Comment by sehrish30** [comment]
the issue also exists in onLoad in iPhone it gets triggered too fast before image is 100% loaded, and my loading image does not appear, 

works fine in web and android

**Comment by marcusmichaels** [comment]
This is still an issue @nextjs-bot 

_Edit by maintainer bot: Comment was **automatically** minimized because it was considered unhelpful. (If you think this was by mistake, let us know). Please only comment if it adds context to the issue. If you want to express that you have the same problem, use the upvote 👍 on the issue description or subscribe to the issue for updates. Thanks!_

**Comment by lorenzrabs** [comment]
The problem still occurs 15.5.12

_Edit by maintainer bot: Comment was **automatically** minimized because it was considered unhelpful. (If you think this was by mistake, let us know). Please only comment if it adds context to the issue. If you want to express that you have the same problem, use the upvote 👍 on the issue description or subscribe to the issue for updates. Thanks!_

**Comment by woohyun-park** [comment]
I opened a PR that removes the built-in blur placeholder after the image has loaded and decoded, without waiting for hydration: #98338.

The change is scoped to the built-in `next/image` placeholder and does not change the timing of user `onLoad` or `onError` callbacks.
